## Question 1
For each of the Theta vectors below, find the decision boundary, and give an example of a point that would get classified as Class 0 and a point that would get classified as Class 1. (There isn't really any Python coding to do here, so you can just type your answers below.)
<ol>
    <li> $\theta =[5,-2]$ </li> 
     <li> $\theta =[5,-2,3]$ </li> 
     <li> $\theta =[5,-2,3,-4]$ </li> 
     <li> $\theta =[5,-2,3,-4,1]$ </li> 
    </ol>

1. $-5 + 2x_1 = 0$, therefore $x_1 = \frac{5}{2}$ \
class0 = 1, class1 = 4 
2. $-5 + 2x_1 - 3x_2 = 0$, therefore $x_2 = \frac{5}{3} - \frac{2}{3}x_1$ \
class0 = (-1,1), class1 = (-3,3)
3. $-5 + 2x_1 - 3x_2 + 4x_3 = 0$, therefore $x_3 = \frac{5}{4} - \frac{2}{4}x_1 + \frac{3}{4}x_2$ \
class0 = (0,0,0), class1 = (-3,2,5)
4. $-5 + 2x_1 - 3x_2 + 4x_3 - 1x_4 = 0$, therefore $x_4 = -5+2x_1 - 3x_2 + 4x_3$ \
class0 = (1,0,0,-1), class1 = (1,0,0,3)

## Question 2

Compare the three different methods of Scaling/Normalizing data from sklearn on the given $X$ vector. Explain what each technique is doing to normalize the data.

In [17]:
## This cell is just set up.
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MaxAbsScaler
from sklearn.preprocessing import MinMaxScaler
X=np.array([1.0,5,-2,4,-6,7,2,-8,9,0])
X=X.reshape(-1,1)

In [18]:
## Example for first method = StandardScaler
scaler1=StandardScaler()
Xss=scaler1.fit(X)
scaler1.transform(X)

array([[-0.03880753],
       [ 0.737343  ],
       [-0.62092042],
       [ 0.54330537],
       [-1.39707095],
       [ 1.12541826],
       [ 0.15523011],
       [-1.78514621],
       [ 1.51349353],
       [-0.23284516]])

(A) What does StandardScaler do? \
    `StandardScalar` essentialy standardizes the data using the mean sample difference divided by the stadard deviation.  By checking `StandardScalar?`, it states the function will remove the mean and scales the data to the unit variance. \
Source "scikit-learn 0.22.1"

(B) Explain what the following code is doing and why it is important. \

In [19]:
print(scaler1.mean_)
print(scaler1.scale_)

[1.2]
[5.15363949]


`scaler1.mean_` is printing the mean of the standardized values which are the same as the mean of the original array X.  `scaler1.scale_` is printing the standard deviation of the scales which is the same as the standard deviation of X.  The code is important because it allows one to access the scales and to see that the scaling is working correctly.

(C)  Repeat (A) and (B) as above for MinMaxScaler in place of StandardScaler and explain what that method is doing. You'll need to think about what the appropriate parameters are for MinMaxScaler for (B).

In [20]:
## Example for second method: MinMaxScaler
scaler2=MinMaxScaler()
Xmms=scaler2.fit(X)
scaler2.transform(X)

array([[0.52941176],
       [0.76470588],
       [0.35294118],
       [0.70588235],
       [0.11764706],
       [0.88235294],
       [0.58823529],
       [0.        ],
       [1.        ],
       [0.47058824]])

(A) What does MinMaxSacler() do? \
This is another normalizing technique that uses the min and max (a range) to scale the data. Based on `MinMaxScaler?` it states that the function will transform features by scaling each feature to a given range.
"This estimator scales and translates each feature individually such that it is in the given range on the training set, e.g. between zero and one." \
source: scikit-learn 0.22.1


(B) Explain what the following code is doing and why it's important \
The following code outputs the raw data's min and max while showing the scale when normalizing the data.  This is useful because we can see if the scaling works and well as it gives us the possibility to save the scaling factor.

In [21]:
print(scaler2.data_min_)
print(scaler2.data_max_)
print(scaler2.scale_)

[-8.]
[9.]
[0.05882353]


In [22]:
## Example for third method: MaxAbsScaler
scaler3=MaxAbsScaler()
Xmas=scaler3.fit(X)
scaler3.transform(X)

array([[ 0.11111111],
       [ 0.55555556],
       [-0.22222222],
       [ 0.44444444],
       [-0.66666667],
       [ 0.77777778],
       [ 0.22222222],
       [-0.88888889],
       [ 1.        ],
       [ 0.        ]])

(A) What does MaxAbsScaler do? \
A scaling technique which will scale each feature by its absolute maximum value.  By doing `MaxAbsScaler?` we see that the estimator scales and translates each feature individually "such that the maximal absolute value of each feature in the training set will be 1.0. It does not shift/center the data, and thus does not destroy any sparsity" \
source: scikit-learn 0.22.1

(B) Explain what the following code is doing and why it's important \
The paramters include the max absolute values from ther raw data as well as the scale factor.  It is worth while to save the scaling incase one needs to rescale later on.

In [23]:
print(scaler3.max_abs_)
print(scaler3.scale_)

[9.]
[9.]


## Question 3
Use sklearn to redo your linear regression from HW due Jan 27, Question 3. That is:
<br>
(A) Create an X which contains only the features you used before.
<br>
(B) Drop NaNs and Normalize the data. (Use one of the sklearn functions to do the Normalization now.)
<br>
(C) Perform the Linear Regression using sklearn commands and output your $\theta$ vector.

In [24]:
# Here's data that you used before.
import pandas as pd
URL = 'http://facweb1.redlands.edu/fac/joanna_bieri/machinelearning/LifeExpectancyData.csv'
df = pd.read_csv(URL)
df.drop('Country', axis=1, inplace=True)
df.drop('Year', axis=1, inplace=True)
df.drop('Status', axis=1, inplace=True)
df.describe()

Life expectancy   Adult Mortality  Infant deaths      Alcohol  \
count       2928.000000      2928.000000    2938.000000  2744.000000   
mean          69.224932       164.796448      30.303948     4.602861   
std            9.523867       124.292079     117.926501     4.052413   
min           36.300000         1.000000       0.000000     0.010000   
25%           63.100000        74.000000       0.000000     0.877500   
50%           72.100000       144.000000       3.000000     3.755000   
75%           75.700000       228.000000      22.000000     7.702500   
max           89.000000       723.000000    1800.000000    17.870000   

       percentage expenditure  Hepatitis B       Measles           BMI  \
count             2938.000000  2385.000000    2938.000000  2904.000000   
mean               738.251295    80.940461    2419.592240    38.321247   
std               1987.914858    25.070016   11467.272489    20.044034   
min                  0.000000     1.000000       0.000000     1.000000   
25%                  4.685343    77.000000       0.000000    19.300000   
50%                 64.912906    92.000000      17.000000    43.500000   
75%                441.534144    97.000000     360.250000    56.200000   
max              19479.911610    99.000000  212183.000000    87.300000   

       under-five deaths         Polio  Total expenditure  Diphtheria   \
count         2938.000000  2919.000000         2712.00000  2919.000000   
mean            42.035739    82.550188            5.93819    82.324084   
std            160.445548    23.428046            2.49832    23.716912   
min              0.000000     3.000000            0.37000     2.000000   
25%              0.000000    78.000000            4.26000    78.000000   
50%              4.000000    93.000000            5.75500    93.000000   
75%             28.000000    97.000000            7.49250    97.000000   
max           2500.000000    99.000000           17.60000    99.000000   

          HIV/AIDS            GDP    Population  thinness  1-19 years  \
count  2938.000000    2490.000000  2.286000e+03           2904.000000   
mean      1.742103    7483.158469  1.275338e+07              4.839704   
std       5.077785   14270.169342  6.101210e+07              4.420195   
min       0.100000       1.681350  3.400000e+01              0.100000   
25%       0.100000     463.935626  1.957932e+05              1.600000   
50%       0.100000    1766.947595  1.386542e+06              3.300000   
75%       0.800000    5910.806335  7.420359e+06              7.200000   
max      50.600000  119172.741800  1.293859e+09             27.700000   

       thinness 5-9 years  Income composition of resources    Schooling  
count         2904.000000                      2771.000000  2775.000000  
mean             4.870317                         0.627551    11.992793  
std              4.508882                         0.210904     3.358920  
min              0.100000                         0.000000     0.000000  
25%              1.500000                         0.493000    10.100000  
50%              3.300000                         0.677000    12.300000  
75%              7.200000                         0.779000    14.300000  
max             28.600000                         0.948000    20.700000

In [25]:
#Create an X which contains only the features you used before. 

df_X = df[['Alcohol', 'Schooling', 'Income composition of resources', 'Life expectancy ']].copy()
print("the shape of df_new is: \n", df_X.shape)
df_X.dropna(inplace=True)
df_X.head()


the shape of df_new is: 
 (2938, 4)


Alcohol  Schooling  Income composition of resources  Life expectancy 
0     0.01       10.1                            0.479              65.0
1     0.01       10.0                            0.476              59.9
2     0.01        9.9                            0.470              59.9
3     0.01        9.8                            0.463              59.5
4     0.01        9.5                            0.454              59.2

In [26]:
# I dont want to normalize my y so I broke it up after removing NA's 

df_data = df_X[['Alcohol', 'Schooling', 'Income composition of resources']].copy()
df_target = df_X[['Life expectancy ']].copy()

In [27]:
# normalize data using StandardScalar

scaler_a = StandardScaler()
Xsc = scaler_a.fit(df_data)
X_norm = scaler_a.transform(df_data)
print(X_norm)

[[-1.12570563 -0.58208997 -0.70116261]
 [-1.12570563 -0.6126342  -0.71543321]
 [-1.12570563 -0.64317844 -0.7439744 ]
 ...
 [-0.02255412 -0.6126342  -0.94851959]
 [-0.69892077 -0.67372268 -0.94851959]
 [-0.70890404 -0.67372268 -0.91522153]]


In [28]:
## Use sklearn package to do Linear Regression
from sklearn.linear_model import LinearRegression
lm = LinearRegression()  ## just a shortcut to type less
X=X_norm  ##Note the double brackets [[]] makes X a DataFrame 
print(X_norm.shape)
print(type(X_norm))
Y=df_target  ## Y is a df
print(Y.shape)
print(type(Y))
lm.fit(X,Y)  ## The .fit command performs the linear regression 

(2584, 3)
<class 'numpy.ndarray'>
(2584, 1)
<class 'pandas.core.frame.DataFrame'>


LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [29]:
theta0 = lm.intercept_ # theta0 is the intercept 
print(theta0)
thetas = lm.coef_
print(type(thetas))
print(thetas)
params = np.hstack((theta0[0], thetas[0,:])) # stack all thetas into one array
params

[69.29732972]
<class 'numpy.ndarray'>
[[-0.45263069  4.79482673  3.11787864]]


array([69.29732972, -0.45263069,  4.79482673,  3.11787864])

## Question 4 
Repeat the above but use the other normalization technique from above. How did the theta parameters change?

In [30]:
# using MaxAbsScaler to normalize
scaler_b = MaxAbsScaler()
X = scaler_b.fit(df_data)
X_norm = scaler_b.transform(df_data)
print(X_norm)

# linear regression from SkiLearn
lm = LinearRegression()  ## just a shortcut to type less
X=X_norm  ##Note the double brackets [[]] makes X a DataFrame 
print(X_norm.shape)
print(type(X_norm))
Y=df_target 
print(Y.shape)
print(type(Y))
lm.fit(X,Y)  ## The .fit command performs the linear regression 

# find thetas
theta0 = lm.intercept_
print(theta0)
thetas = lm.coef_
print(thetas)
params=np.hstack((theta0, thetas[0]))
params

[[0.0005596  0.48792271 0.50527426]
 [0.0005596  0.48309179 0.5021097 ]
 [0.0005596  0.47826087 0.49578059]
 ...
 [0.24790151 0.48309179 0.45042194]
 [0.0962507  0.47342995 0.45042194]
 [0.09401231 0.47342995 0.45780591]]
(2584, 3)
<class 'numpy.ndarray'>
(2584, 1)
<class 'pandas.core.frame.DataFrame'>
[42.93473474]
[[-2.01874491 30.31604634 14.06009828]]


array([42.93473474, -2.01874491, 30.31604634, 14.06009828])

My $\theta$ changed quite a bit from the $\theta$s I recieved from the StandardScaler. They went from: \
array([69.29732972, -0.45263069,  4.79482673,  3.11787864]) to \
array([42.93473474, -2.01874491, 30.31604634, 14.06009828]). \
The $\theta$s remained in the same "order", where $\theta_0$ was the biggest and $\theta_1$ was the smallest.

## Question 5
Repeat the above but do not normalize the data before calling the LinearRegression() function. Instead call the LinearRegression() function with 
lm = LinearRegression(normalize=True). How did the theta parameters change? What is this doing?

In [31]:
X = df_data

lm = LinearRegression(normalize = True)  ## just a shortcut to type less
Y = df_target  ## Y is still a Series, note shape.
print(Y.shape)
print(type(Y))
lm.fit(X,Y)  ## The .fit command performs the linear regression 

# find thetas
theta0 = lm.intercept_
print(theta0)
thetas = lm.coef_
print(thetas)
params=np.hstack((theta0, thetas[0]))
params

(2584, 1)
<class 'pandas.core.frame.DataFrame'>
[42.93473474]
[[-0.11296838  1.4645433  14.8313273 ]]


array([42.93473474, -0.11296838,  1.4645433 , 14.8313273 ])

The $\theta$s followed the same pattern where $\theta_0$ is the largest and $\theta_1$ is the smallest.  However, the actual values seem to be a compromise between the AbsMx method and the StandardScaler.

## Question 6
For your favorite normalization technique, pick a different set of 3-5 features and see how the theta parameters change. 

In [32]:
df_X = df[['Infant deaths', 'Total expenditure', 'Income composition of resources', 'Diphtheria ', 'Life expectancy ']].copy()
print("the shape of df_new is: \n", df_X.shape)
df_X.dropna(inplace=True)
df_X.head()

df_data = df_X[['Infant deaths', 'Total expenditure', 'Income composition of resources', 'Diphtheria ']].copy()
df_target = df_X[['Life expectancy ']].copy()

X = df_data

lm = LinearRegression(normalize = True)  ## just a shortcut to type less
Y=df_target  ## Y is still a Series, note shape.
print(Y.shape)
print(type(Y))
lm.fit(X,Y)  ## The .fit command performs the linear regression 

# find thetas
theta0 = lm.intercept_
print(theta0)
thetas = lm.coef_
print(thetas)
params=np.hstack((theta0[0], thetas[0]))
params


the shape of df_new is: 
 (2938, 5)
(2574, 1)
<class 'pandas.core.frame.DataFrame'>
[43.97512022]
[[-3.66440079e-03  1.97591441e-01  2.81245167e+01  7.99047782e-02]]


array([ 4.39751202e+01, -3.66440079e-03,  1.97591441e-01,  2.81245167e+01,
        7.99047782e-02])

The $\theta$ values seem pretty similar despite it being output in scienfitific notation.